File: 0994-rotting-oranges/0994-rotting-oranges.cs
public class Solution {
    public int OrangesRotting(int[][] grid) {
        int m = grid.Length, n = grid[0].Length, fresh = 0;
        var rottenIdx = new Queue<int[]>();
        
        for(var i=0; i<m; i++){
            for(var j=0; j<n; j++){
                if(grid[i][j] == 1)
                    fresh++;
                else if(grid[i][j] == 2)
                    rottenIdx.Enqueue([i, j]);
            }
        }

        if(fresh == 0) return 0;    // nothing is fresh
        if(rottenIdx.Count == 0) return -1;    // no rotten, no conversion possible
        
        (int, int)[] dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)];
        for(var minute=1; rottenIdx.Count > 0; minute++){
            for(var i=rottenIdx.Count; i>0; i--){
                var point = rottenIdx.Dequeue();
                int x = point[0], y = point[1];

                foreach(var (dx, dy) in dirs){
                    int nx = x + dx, ny = y + dy;
                    if(nx == -1 || ny == -1 || nx == m || ny == n || grid[nx][ny] != 1) continue;
                    
                    fresh--;
                    grid[nx][ny] = 0;
                    rottenIdx.Enqueue([nx, ny]);
                }
            }
            if(fresh == 0) return minute;
        }

        return -1;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m, n, fresh = len(grid), len(grid[0]), 0
        rottenIdx = deque()
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh += 1
                elif grid[i][j] == 2:
                    rottenIdx.append((i, j))

        if fresh == 0: return 0  # nothing is fresh
        if not rottenIdx: return -1  # no rotten, no conversion possible
        
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        minute = 1
        while rottenIdx:
            for _ in range(len(rottenIdx)):
                x, y = rottenIdx.popleft()

                for dx, dy in dirs:
                    nx, ny = x + dx, y + dy
                    if nx < 0 or ny < 0 or nx >= m or ny >= n or grid[nx][ny] != 1: continue
                    
                    fresh -= 1
                    grid[nx][ny] = 0
                    rottenIdx.append((nx, ny))

            if fresh == 0: return minute
            minute += 1

        return -1